### DISCUSSION TITLE
Get data from async functions

### DISCUSSION THREAD

**Discussion by ghost** [alpinejs/alpine#1042] [1. Help] [answered]
I use a library that exports async functions to transform some data (internationalization stuff).

Is there a way i can use these functions within Alpine?

```
  <body x-data="{ key: 'key' }">
    <label x-text="localize(key)"></label>
    <label x-text="await localize(key)"></label>
    <label x-text="(await localize(key))"></label>

    <script>
      async function localize(key) {
        return 'value';
      }
    </script>
  </body>
```

- First \<label> shows "[object Promise]"
- Second \<label> throws Alpine Error: "SyntaxError: await is only valid in async functions and async generators"
- Third \<label> throws Alpine Error: "SyntaxError: missing ) in parenthetical"

**Comment by ryangjchandler** [comment]
@fabianobonin You can only use `async/await` in action-based expressions, i.e. event handlers. All expressions that use the return value are synchronous, i.e. `x-text`, `x-bind`.

**Reply by ghost** [reply]
Hi, Ryan, thanks for your fast answer!

I understand and agree that expressions with return values must be synchronous, as we need the return value to fill the component.

So my point is:

Is it possible to Alpine, given it runs async, always await for an expression, so if the expression returns a promise (async) it will evaluate just like a regular expression?

I mean, in javascript there is no harm in await for a sync function, so in case we pass an async expression, it will evaluate it also.

```javascript
        f1 = () => 1;
        f2 = async () => 2;
        console.log(await f1()); // await sync function, will output 1
        console.log(await f2()); // await async function, will output 2
```

Best regards

**Reply by ghost** [reply]
Looking over Alpine source code (without so much compromise), I think if we add 2 "await" in utils.js.saferEval would do it.
What do you think about it?

```javascript
export function saferEval(el, expression, dataContext, additionalHelperVariables = {}) {
    return tryCatch(() => {
        if (typeof expression === 'function') {
            return await expression.call(dataContext)
        }

        return await (new Function(['$data', ...Object.keys(additionalHelperVariables)], `var __alpine_result; with($data) { __alpine_result = ${expression} }; return __alpine_result`))(
            dataContext, ...Object.values(additionalHelperVariables)
        )
    }, { el, expression })
}
```

Best regards

**Reply by ryangjchandler** [reply]
> Looking over Alpine source code (without so much compromise), I think if we add 2 "await" in utils.js.saferEval would do it.
> What do you think about it?
> 
> ```js
> export function saferEval(el, expression, dataContext, additionalHelperVariables = {}) {
>     return tryCatch(() => {
>         if (typeof expression === 'function') {
>             return await expression.call(dataContext)
>         }
> 
>         return await (new Function(['$data', ...Object.keys(additionalHelperVariables)], `var __alpine_result; with($data) { __alpine_result = ${expression} }; return __alpine_result`))(
>             dataContext, ...Object.values(additionalHelperVariables)
>         )
>     }, { el, expression })
> }
> ```
> 
> Best regards

This technically wouldn't work because top level `await` isn't supported in the browser. The `saferEvalNoReturn` function tackles this by providing a custom `AsyncFunction` implemented on top of `Function`.

I'm not sure but couldn't you make the callback passed to `tryCatch` `async`, @HugoDF?

**Reply by HugoDF** [reply]
`tryCatch` will work for a Promise or synchronous error

Dunno if Caleb had a reason for supporting await only in directives where the return value isn't used

**Reply by ryangjchandler** [reply]
Yeah, I think it was initially for Livewire. `x-on:click="name = await $wire.getName()"`.

**Reply by ghost** [reply]
Hi,

Is it doable? Can I fill a feature request somewhere?

**Reply by HugoDF** [reply]
> Hi,
> 
> Is it doable? Can I fill a feature request somewhere?

You can create an issue, there might be a limitation of async functions in these scenarios

**Comment by SimoTod** [comment]
@fabianobonin 

It's not exactly the same thing but you can define empty variables and populate them inside x-init when your promises resolve.

See https://codepen.io/SimoTod/pen/VwmZoyG

**Reply by ghost** [reply]
Hi, Simone, thanks for your answer.

Off course there are other ways to achieve this, but it would be great if we do not need to use workarounds.

Best regards

**Comment by ghost** [comment, marked as answer]
It will be implemented in v3.
